File: app/models/api_auth_config.rb
# frozen_string_literal: true

class ApiAuthConfigValidator < ActiveModel::Validator
  def validate (auth_config)
    # The combination of [org id, service entity id, idp entity id]
    # must be unique, unless the entity ids haven't been set yet.
    return if auth_config.org.nil? ||
      auth_config.service_entity_id.blank? ||
      auth_config.idp_entity_id.blank?

    existing_count = ApiAuthConfig.where(
      :org_id => auth_config.org.id,
      :service_entity_id => auth_config.service_entity_id,
      :idp_entity_id => auth_config.idp_entity_id
    ).where.not(:id => auth_config.id).count()

    if (existing_count > 0)
      auth_config.errors.add :base, "service_entity_id/idp_entity_id combination already in use"
    end
  end
end

class ApiAuthConfig < ApplicationRecord

  class SecretConfigMarshaler
    def dump (value)
      value.to_json
    end
    def load (value_json)
      begin
        JSON.parse(value_json)
      rescue
        {}
      end
    end
  end

  self.primary_key = :id

  include Api::V1::Schema
  include JsonAccessor

  Protocols = {
    :saml => "saml",
    :oidc => "oidc",
    :password => "password",
    :google => "google",
    :omni => "omni"
  }.freeze
  Default_Idp_Protocol = Protocols[:saml]

  Default_Name_Identifier_Format = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
  Default_Security_Settings = {
    :authn_requests_signed => false,
    :logout_requests_signed => false,
    :logout_responses_signed => false,
    :metadata_signed => false,
    :digest_method => XMLSecurity::Document::SHA1,
    :signature_method => XMLSecurity::Document::RSA_SHA1
  }

  validates_with ApiAuthConfigValidator
  belongs_to :owner, class_name: "User", foreign_key: "owner_id", required: true
  belongs_to :org

  json_accessor :security_settings, :oidc_id_claims, :oidc_access_claims, :client_config

  attr_encrypted :secret_config,
    prefix: "", suffix: "_enc",
    marshal: true, marshaler: SecretConfigMarshaler.new, dump_method: :dump, load_method: :load,
    key: :enc_key, encode: "m0", encode_iv: "m0",
    encode_salt: true

  def enc_key
    API_SECRETS[:enc][:auth_config_key][0..31]
  end

  before_save  do
    self.uid = ApiAuthConfig.generate_uid if self.uid.blank?
    self.security_settings = Default_Security_Settings if self.security_settings.blank?
  end

  def self.generate_uid
    tmp_uid = nil

    1.upto(5) do
      tmp_uid = rand(36**8).to_s(36)
      break if ApiAuthConfig.select(:uid).find_by_uid(tmp_uid).nil?
    end

    if (tmp_uid.nil?)
      raise Api::V1::ApiError.new(:internal_server_error, "Could not generate org idp uid")
    end

    tmp_uid
  end

  def self.generate_base_url (request)
    proto = (Rails.env.development? ? "http" : "https") + 
      "://" + request.env["SERVER_NAME"]

    port = request.env["SERVER_PORT"]
    proto += ":#{port}" if (!port.blank? && (port != "80"))

    context = ""
    if (Rails.env.qa?)
      context = "admin-api"
    elsif (!Rails.env.development?)
      context = "nexla-api"
    end
    
    proto += "/#{context}" if !context.blank?
    return proto
  end

  def self.sso_options
    # BEWARE this method is called from an unauthenticated route.
    # Be careful about what attributes are revealed.
    ApiAuthConfig.where(global: true).map(&:public_attributes) || []
  end

  def self.get_mapping_config(uid)
    mapping = ApiAuthConfig.find_by(uid: uid)
    return mapping if mapping.nil?
    return mapping.secret_config.is_a?(Hash) ?
      mapping.client_config.deep_merge(mapping.secret_config) : mapping.client_config
  end

  def self.build_from_input (api_user_info, input, request)
    if (input.blank? || api_user_info.nil? || api_user_info.input_owner.nil?)
      raise Api::V1::ApiError.new(:bad_request, "Required input missing")
    end

    mapping_org = api_user_info.input_org
    if (mapping_org.nil?)
      raise Api::V1::ApiError.new(:bad_request, "Required input org missing")
    end

    if (!mapping_org.has_admin_access?(api_user_info.user))
      raise Api::V1::ApiError.new(:forbidden, "IDP mappings can only be created by org admins")
    end

    if (input.key?(:global) && input[:global] && !api_user_info.user.super_user?)
      raise Api::V1::ApiError.new(:forbidden, "Global mappings can only be created by Nexla administrators.")
    end

    if (input.key?(:uid) && !input.key?(:global))
      raise Api::V1::ApiError.new(:bad_request, "UID can only be specified for global mappings")
    end

    protocol = (input[:protocol] || Default_Idp_Protocol).downcase
    if (Protocols.key?(protocol))
      raise Api::V1::ApiError.new(:bad_request, "Unsupported IDP protocol")
    end

    if (input[:name].blank?)
      raise Api::V1::ApiError.new(:bad_request, "Mapping name must not be blank")
    end

    auth_config = ApiAuthConfig.new
    auth_config.uid = ApiAuthConfig.generate_uid
    auth_config.protocol = protocol
    auth_config.update_mutable!(api_user_info, input, request)

    return auth_config
  end

  def update_mutable! (api_user_info, input, request)
    self.update_mapping_owner(api_user_info, input)
    self.name = input[:name] if !input[:name].blank?
    self.description = input[:description] if input.key?(:description)

    if (input.key?(:global) && input[:global] && !api_user_info.user.super_user?)
      raise Api::V1::ApiError.new(:forbidden, "Global mappings can only be created by Nexla administrators.")
    end
    self.global = input[:global] if input.key?(:global)

    if input.key?(:uid)
      if !input.key?(:global)
        # we check this here as well as in build_from_input to check for both update and
        # create calls while still being able to error out before creating a new mapping
        #
        # Custom UIDs are only allowed for global mappings, since they can be used to define the path for authenticating
        # via a specif auth method.
        #
        # For example, if a global mapping with UID "myidp" is created, then the path for authenticating via that IDP
        # would be /token/myidp
        raise Api::V1::ApiError.new(:bad_request, "UID can only be specified for global mappings")
      elsif ApiAuthConfig.find_by_uid(input[:uid]).present?
        raise Api::V1::ApiError.new(:bad_request, "UID already in use")
      else
        self.uid = input[:uid]
      end
    end

    self.auto_create_users_enabled = input[:auto_create_users_enabled] if input.key?(:auto_create_users_enabled)

    if (!input[:nexla_base_url].blank? || self.nexla_base_url.blank?)
      base_url = input[:nexla_base_url].blank? ?
        ApiAuthConfig.generate_base_url(request) : input[:nexla_base_url]
      while ("/" == base_url.last) do
        base_url.chomp!("/")
      end
      self.nexla_base_url = base_url if !base_url.blank?
    end

    if self.is_saml?
      self.name_identifier_format = (input[:name_identifier_format] || 
        Default_Name_Identifier_Format)

      if (!input[:service_entity_id].blank? || self.service_entity_id.blank?)
        self.service_entity_id = input[:service_entity_id].blank? ?
          self.nexla_base_url : input[:service_entity_id]
      end

      self.assertion_consumer_url = input[:assertion_consumer_url] if input.key?(:assertion_consumer_url)
      self.idp_entity_id = input[:idp_entity_id] if input.key?(:idp_entity_id)
      self.idp_sso_target_url = input[:idp_sso_target_url] if input.key?(:idp_sso_target_url)
      self.idp_slo_target_url = input[:idp_slo_target_url] if input.key?(:idp_slo_target_url)
      self.idp_cert = input[:idp_cert] if input.key?(:idp_cert)
    elsif self.is_oidc?
      self.oidc_domain = input[:oidc_domain] if input.key?(:oidc_domain)
      self.oidc_keys_url_key = input[:oidc_keys_url_key] if input.key?(:oidc_keys_url_key)
      self.oidc_id_claims = input[:oidc_id_claims] if input[:oidc_id_claims].is_a?(Hash)
      self.oidc_access_claims = input[:oidc_access_claims] if input[:oidc_access_claims].is_a?(Hash)
    end

    self.security_settings = input[:security_settings] if input.key?(:security_settings)
    self.metadata = input[:metadata] if input.key?(:metadata)
    self.client_config = input[:client_config] if input.key?(:client_config)
    self.secret_config = input[:secret_config] || {}
    self.save!
  end

  Protocols.keys.each do |protocol_name|
    define_method "is_#{protocol_name}?" do
      self.protocol == Protocols[protocol_name]
    end
  end

  def public_attributes

    # BEWARE this method is called from an unauthenticated route.
    # Be careful about what attributes are revealed.
    load_public_attribute_config

    attrs = {}

    if @@auth_config_spec.present?
      keys = @@auth_config_spec["COMMON"] + (@@auth_config_spec[self.protocol] || [])
      keys.each do |key|
        if self.respond_to?(key)
          attrs[key] = self.public_send(key)
          if (key == "client_config")
            attrs.dig(key, "web")&.delete("client_secret")
          end
        end
      end
    end

    return attrs
  end

  def load_public_attribute_config
    @@auth_config_spec ||= nil
    return if !@@auth_config_spec.nil?
    config_spec = JSON.parse(File.read("#{Rails.root}/config/AuthConfigPublicAttributes.json"))
    @@auth_config_spec = config_spec if config_spec["COMMON"].present?
  end

  def assertion_consumer_url
    return nil if !self.is_saml?
    acu = super
    return (acu.blank? ? "#{self.nexla_base_url}/token/#{self.uid}" : acu)
  end

  def logout_url
    "#{self.nexla_base_url}/logout/#{self.uid}"
  end

  def metadata_url
    "#{self.nexla_base_url}/metadata/#{self.uid}"
  end

  def oidc_token_verify_url
    self.is_oidc? ? "#{self.nexla_base_url}/token/#{self.uid}" : nil
  end

  def get_assertion_settings
    settings = OneLogin::RubySaml::Settings.new

    settings.soft = true
    settings.issuer = self.service_entity_id
    settings.assertion_consumer_service_url = self.assertion_consumer_url
    settings.assertion_consumer_logout_service_url = self.logout_url

    settings.idp_entity_id = self.idp_entity_id
    settings.idp_sso_target_url = self.idp_sso_target_url
    settings.idp_slo_target_url = self.idp_slo_target_url
    settings.idp_cert = self.idp_cert
    settings.name_identifier_format = self.name_identifier_format

    ss = self.security_settings
    ss = Default_Security_Settings if ss.blank?

    ss.keys.each do |key|
      settings.security[key] = ss[key]
    end

    return settings
  end

  protected

  def update_mapping_owner (api_user_info, input)
    mapping_owner = api_user_info.input_owner
    raise Api::V1::ApiError.new(:not_found, "User not found") if mapping_owner.nil?

    mapping_org = api_user_info.input_org
    raise Api::V1::ApiError.new(:bad_request, "Input org_id required") if mapping_org.nil?

    return if (mapping_owner == self.owner) && (mapping_org == self.org)

    if (!mapping_owner.org_member?(mapping_org))
      raise Api::V1::ApiError.new(:bad_request, "Mapping owner must be an org member")
    end

    if (!mapping_org.has_admin_access?(mapping_owner))
      raise Api::V1::ApiError.new(:bad_request, "Mapping owner must be an org admin")
    end

    self.owner = mapping_owner
    self.org = mapping_org
  end

end

File: app/models/api_auth_config.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Boolean, Enum as SQLEnum, JSON
from sqlalchemy.orm import relationship, validates
from datetime import datetime
from enum import Enum as PyEnum
from typing import Optional, Dict, Any, Union
import json
import hashlib
import random
import string

from app.database import Base

class Protocols(PyEnum):
    SAML = "saml"
    OIDC = "oidc"
    PASSWORD = "password"
    GOOGLE = "google"
    OMNI = "omni"

class ApiAuthConfig(Base):
    __tablename__ = 'api_auth_configs'
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    owner_id = Column(Integer, ForeignKey('users.id'), nullable=False)
    org_id = Column(Integer, ForeignKey('orgs.id'), nullable=False)
    
    uid = Column(String(255), unique=True)
    name = Column(String(255), nullable=False)
    description = Column(Text)
    protocol = Column(SQLEnum(Protocols), nullable=False, default=Protocols.SAML)
    
    # Global configuration flag
    global_ = Column('global', Boolean, default=False)
    
    # User auto-creation settings
    auto_create_users_enabled = Column(Boolean, default=False)
    
    # Base URL configuration
    nexla_base_url = Column(Text)
    
    # SAML-specific fields
    name_identifier_format = Column(Text)
    service_entity_id = Column(Text)
    assertion_consumer_url = Column(Text)
    idp_entity_id = Column(Text)
    idp_sso_target_url = Column(Text)
    idp_slo_target_url = Column(Text)
    idp_cert = Column(Text)
    
    # OIDC-specific fields
    oidc_domain = Column(Text)
    oidc_keys_url_key = Column(Text)
    oidc_id_claims = Column(JSON)
    oidc_access_claims = Column(JSON)
    
    # Configuration storage
    security_settings = Column(JSON)
    extra_metadata = Column(Text)
    client_config = Column(JSON)
    secret_config = Column(Text)  # Encrypted in production
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    owner = relationship("User", foreign_keys=[owner_id], back_populates="api_auth_configs")
    org = relationship("Org", foreign_keys=[org_id], back_populates="api_auth_configs")
    
    # Constants
    DEFAULT_IDP_PROTOCOL = Protocols.SAML
    DEFAULT_NAME_IDENTIFIER_FORMAT = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
    DEFAULT_SECURITY_SETTINGS = {
        "authn_requests_signed": False,
        "logout_requests_signed": False,
        "logout_responses_signed": False,
        "metadata_signed": False,
        "digest_method": "SHA1",
        "signature_method": "RSA_SHA1"
    }
    
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        if not self.uid:
            self.uid = self.generate_uid()
        if not self.security_settings:
            self.security_settings = self.DEFAULT_SECURITY_SETTINGS.copy()
    
    @classmethod
    def generate_uid(cls, max_attempts: int = 5) -> str:
        """Generate unique UID for auth config"""
        for attempt in range(max_attempts):
            # Generate 8 character alphanumeric string
            tmp_uid = ''.join(random.choices(string.ascii_lowercase + string.digits, k=8))
            
            # Check if UID already exists
            existing = cls.query.filter_by(uid=tmp_uid).first()
            if not existing:
                return tmp_uid
        
        raise ValueError("Could not generate unique auth config UID")
    
    @classmethod
    def generate_base_url(cls, request_host: str = None, request_port: str = None, 
                         is_development: bool = False, is_qa: bool = False) -> str:
        """Generate base URL for auth configuration"""
        proto = "http" if is_development else "https"
        
        if request_host:
            base_url = f"{proto}://{request_host}"
            
            if request_port and request_port not in ["80", "443"]:
                base_url += f":{request_port}"
        else:
            # Default fallback
            base_url = f"{proto}://localhost"
        
        # Add context path based on environment
        context = ""
        if is_qa:
            context = "admin-api"
        elif not is_development:
            context = "nexla-api"
        
        if context:
            base_url += f"/{context}"
        
        return base_url.rstrip('/')
    
    @classmethod
    def sso_options(cls) -> list:
        """Get public SSO options for unauthenticated routes"""
        global_configs = cls.query.filter_by(global_=True).all()
        return [config.public_attributes() for config in global_configs]
    
    @classmethod
    def get_mapping_config(cls, uid: str) -> Optional[Dict[str, Any]]:
        """Get merged mapping configuration by UID"""
        mapping = cls.query.filter_by(uid=uid).first()
        if not mapping:
            return None
        
        config = mapping.client_config or {}
        
        # Merge with secret config if available
        if mapping.secret_config:
            try:
                secret_config = json.loads(mapping.secret_config) if isinstance(mapping.secret_config, str) else mapping.secret_config
                if isinstance(secret_config, dict):
                    config.update(secret_config)
            except (json.JSONDecodeError, TypeError):
                pass
        
        return config
    
    @classmethod
    def build_from_input(cls, user, org, input_data: Dict[str, Any], 
                        request_info: Dict[str, Any] = None) -> 'ApiAuthConfig':
        """Factory method to create auth config from input data"""
        if not input_data or not user or not org:
            raise ValueError("Required input missing")
        
        if not org.has_admin_access_(user):
            raise ValueError("Auth configs can only be created by org admins")
        
        if input_data.get('global') and not user.super_user_:
            raise ValueError("Global mappings can only be created by Nexla administrators")
        
        if input_data.get('uid') and not input_data.get('global'):
            raise ValueError("UID can only be specified for global mappings")
        
        protocol = (input_data.get('protocol', cls.DEFAULT_IDP_PROTOCOL.value)).lower()
        try:
            protocol_enum = Protocols(protocol)
        except ValueError:
            raise ValueError("Unsupported IDP protocol")
        
        if not input_data.get('name'):
            raise ValueError("Mapping name must not be blank")
        
        auth_config = cls(protocol=protocol_enum)
        auth_config.update_mutable_(user, org, input_data, request_info)
        
        return auth_config
    
    def update_mutable_(self, user, org, input_data: Dict[str, Any], 
                       request_info: Dict[str, Any] = None) -> None:
        """Update mutable attributes"""
        self._update_mapping_owner(user, org)
        
        # Basic attributes
        if input_data.get('name'):
            self.name = input_data['name']
        if 'description' in input_data:
            self.description = input_data['description']
        
        # Global flag (admin only)
        if 'global' in input_data:
            if input_data['global'] and not user.super_user_:
                raise ValueError("Global mappings can only be created by Nexla administrators")
            self.global_ = input_data['global']
        
        # UID handling (global mappings only)
        if 'uid' in input_data:
            if not input_data.get('global'):
                raise ValueError("UID can only be specified for global mappings")
            
            existing = self.__class__.query.filter_by(uid=input_data['uid']).first()
            if existing and existing.id != self.id:
                raise ValueError("UID already in use")
            
            self.uid = input_data['uid']
        
        if 'auto_create_users_enabled' in input_data:
            self.auto_create_users_enabled = input_data['auto_create_users_enabled']
        
        # Base URL handling
        if 'nexla_base_url' in input_data or not self.nexla_base_url:
            if input_data.get('nexla_base_url'):
                base_url = input_data['nexla_base_url'].rstrip('/')
            else:
                # Generate from request info
                base_url = self.generate_base_url(
                    request_info.get('host') if request_info else None,
                    request_info.get('port') if request_info else None,
                    request_info.get('is_development', False) if request_info else False,
                    request_info.get('is_qa', False) if request_info else False
                )
            self.nexla_base_url = base_url
        
        # Protocol-specific configuration
        if self.is_saml_():
            self._update_saml_config(input_data)
        elif self.is_oidc_():
            self._update_oidc_config(input_data)
        
        # General configuration
        if 'security_settings' in input_data:
            self.security_settings = input_data['security_settings']
        if 'metadata' in input_data:
            self.extra_metadata = input_data['metadata']
        if 'client_config' in input_data:
            self.client_config = input_data['client_config']
        if 'secret_config' in input_data:
            self.secret_config = json.dumps(input_data['secret_config']) if isinstance(input_data['secret_config'], dict) else input_data['secret_config']
    
    def _update_saml_config(self, input_data: Dict[str, Any]) -> None:
        """Update SAML-specific configuration"""
        self.name_identifier_format = input_data.get(
            'name_identifier_format', self.DEFAULT_NAME_IDENTIFIER_FORMAT)
        
        if 'service_entity_id' in input_data or not self.service_entity_id:
            self.service_entity_id = input_data.get('service_entity_id') or self.nexla_base_url
        
        if 'assertion_consumer_url' in input_data:
            self.assertion_consumer_url = input_data['assertion_consumer_url']
        if 'idp_entity_id' in input_data:
            self.idp_entity_id = input_data['idp_entity_id']
        if 'idp_sso_target_url' in input_data:
            self.idp_sso_target_url = input_data['idp_sso_target_url']
        if 'idp_slo_target_url' in input_data:
            self.idp_slo_target_url = input_data['idp_slo_target_url']
        if 'idp_cert' in input_data:
            self.idp_cert = input_data['idp_cert']
    
    def _update_oidc_config(self, input_data: Dict[str, Any]) -> None:
        """Update OIDC-specific configuration"""
        if 'oidc_domain' in input_data:
            self.oidc_domain = input_data['oidc_domain']
        if 'oidc_keys_url_key' in input_data:
            self.oidc_keys_url_key = input_data['oidc_keys_url_key']
        if 'oidc_id_claims' in input_data and isinstance(input_data['oidc_id_claims'], dict):
            self.oidc_id_claims = input_data['oidc_id_claims']
        if 'oidc_access_claims' in input_data and isinstance(input_data['oidc_access_claims'], dict):
            self.oidc_access_claims = input_data['oidc_access_claims']
    
    def _update_mapping_owner(self, user, org) -> None:
        """Update mapping ownership with validation"""
        if not user:
            raise ValueError("User not found")
        if not org:
            raise ValueError("Input org_id required")
        
        # Skip if already correct
        if user.id == self.owner_id and org.id == self.org_id:
            return
        
        if not user.org_member_(org):
            raise ValueError("Mapping owner must be an org member")
        
        if not org.has_admin_access_(user):
            raise ValueError("Mapping owner must be an org admin")
        
        self.owner_id = user.id
        self.org_id = org.id
    
    @validates('name')
    def validate_name(self, key, name):
        if not name or not name.strip():
            raise ValueError("Name cannot be blank")
        return name.strip()
    
    @validates('uid')
    def validate_uid_uniqueness(self, key, uid):
        if uid:
            existing = self.__class__.query.filter_by(uid=uid).first()
            if existing and existing.id != self.id:
                raise ValueError("UID must be unique")
        return uid
    
    # Protocol predicate methods (Rails pattern)
    def is_saml_(self) -> bool:
        """Check if protocol is SAML"""
        return self.protocol == Protocols.SAML
    
    def is_oidc_(self) -> bool:
        """Check if protocol is OIDC"""
        return self.protocol == Protocols.OIDC
    
    def is_password_(self) -> bool:
        """Check if protocol is password"""
        return self.protocol == Protocols.PASSWORD
    
    def is_google_(self) -> bool:
        """Check if protocol is Google"""
        return self.protocol == Protocols.GOOGLE
    
    def is_omni_(self) -> bool:
        """Check if protocol is Omni"""
        return self.protocol == Protocols.OMNI
    
    def global_(self) -> bool:
        """Check if config is global"""
        return self.global_ is True
    
    def local_(self) -> bool:
        """Check if config is local (not global)"""
        return not self.global_
    
    def auto_create_enabled_(self) -> bool:
        """Check if auto user creation is enabled"""
        return self.auto_create_users_enabled is True
    
    def has_secret_config_(self) -> bool:
        """Check if secret config is present"""
        return bool(self.secret_config)
    
    def complete_saml_config_(self) -> bool:
        """Check if SAML configuration is complete"""
        if not self.is_saml_():
            return False
        
        required_fields = [
            self.service_entity_id,
            self.idp_entity_id,
            self.idp_sso_target_url,
            self.nexla_base_url
        ]
        
        return all(field for field in required_fields)
    
    def complete_oidc_config_(self) -> bool:
        """Check if OIDC configuration is complete"""
        if not self.is_oidc_():
            return False
        
        return bool(self.oidc_domain and self.client_config)
    
    # URL generation methods
    def assertion_consumer_url_computed(self) -> Optional[str]:
        """Get computed assertion consumer URL"""
        if not self.is_saml_():
            return None
        
        if self.assertion_consumer_url:
            return self.assertion_consumer_url
        
        if self.nexla_base_url and self.uid:
            return f"{self.nexla_base_url}/token/{self.uid}"
        
        return None
    
    def logout_url(self) -> Optional[str]:
        """Get logout URL"""
        if not self.nexla_base_url or not self.uid:
            return None
        return f"{self.nexla_base_url}/logout/{self.uid}"
    
    def metadata_url(self) -> Optional[str]:
        """Get metadata URL"""
        if not self.nexla_base_url or not self.uid:
            return None
        return f"{self.nexla_base_url}/metadata/{self.uid}"
    
    def oidc_token_verify_url(self) -> Optional[str]:
        """Get OIDC token verification URL"""
        if not self.is_oidc_() or not self.nexla_base_url or not self.uid:
            return None
        return f"{self.nexla_base_url}/token/{self.uid}"
    
    def public_attributes(self) -> Dict[str, Any]:
        """Get public attributes (safe for unauthenticated routes)"""
        # Base public attributes
        attrs = {
            'uid': self.uid,
            'name': self.name,
            'protocol': self.protocol.value,
            'global': self.global_
        }
        
        # Protocol-specific public attributes
        if self.is_saml_():
            attrs.update({
                'service_entity_id': self.service_entity_id,
                'assertion_consumer_url': self.assertion_consumer_url_computed(),
                'name_identifier_format': self.name_identifier_format,
                'metadata_url': self.extra_metadata_url()
            })
        
        elif self.is_oidc_():
            attrs.update({
                'oidc_domain': self.oidc_domain,
                'oidc_token_verify_url': self.oidc_token_verify_url()
            })
            
            # Include public client config (exclude secrets)
            if self.client_config:
                public_config = self.client_config.copy()
                # Remove sensitive information
                if 'web' in public_config:
                    public_config['web'].pop('client_secret', None)
                attrs['client_config'] = public_config
        
        return attrs
    
    def get_assertion_settings(self) -> Dict[str, Any]:
        """Get SAML assertion settings (for SAML protocol)"""
        if not self.is_saml_():
            return {}
        
        settings = {
            'soft': True,
            'issuer': self.service_entity_id,
            'assertion_consumer_service_url': self.assertion_consumer_url_computed(),
            'assertion_consumer_logout_service_url': self.logout_url(),
            'idp_entity_id': self.idp_entity_id,
            'idp_sso_target_url': self.idp_sso_target_url,
            'idp_slo_target_url': self.idp_slo_target_url,
            'idp_cert': self.idp_cert,
            'name_identifier_format': self.name_identifier_format
        }
        
        # Add security settings
        security_settings = self.security_settings or self.DEFAULT_SECURITY_SETTINGS
        settings['security'] = security_settings
        
        return settings
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API responses"""
        return {
            'id': self.id,
            'uid': self.uid,
            'owner_id': self.owner_id,
            'org_id': self.org_id,
            'name': self.name,
            'description': self.description,
            'protocol': self.protocol.value if self.protocol else None,
            'global': self.global_,
            'auto_create_users_enabled': self.auto_create_users_enabled,
            'nexla_base_url': self.nexla_base_url,
            
            # SAML fields
            'name_identifier_format': self.name_identifier_format if self.is_saml_() else None,
            'service_entity_id': self.service_entity_id if self.is_saml_() else None,
            'assertion_consumer_url': self.assertion_consumer_url_computed() if self.is_saml_() else None,
            'idp_entity_id': self.idp_entity_id if self.is_saml_() else None,
            'idp_sso_target_url': self.idp_sso_target_url if self.is_saml_() else None,
            'idp_slo_target_url': self.idp_slo_target_url if self.is_saml_() else None,
            'logout_url': self.logout_url(),
            'metadata_url': self.extra_metadata_url(),
            
            # OIDC fields
            'oidc_domain': self.oidc_domain if self.is_oidc_() else None,
            'oidc_keys_url_key': self.oidc_keys_url_key if self.is_oidc_() else None,
            'oidc_token_verify_url': self.oidc_token_verify_url() if self.is_oidc_() else None,
            
            # Configuration
            'security_settings': self.security_settings,
            'client_config': self.client_config,
            
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            
            # Predicate methods
            'is_saml': self.is_saml_(),
            'is_oidc': self.is_oidc_(),
            'is_password': self.is_password_(),
            'is_google': self.is_google_(),
            'is_omni': self.is_omni_(),
            'is_global': self.global_(),
            'is_local': self.local_(),
            'auto_create_enabled': self.auto_create_enabled_(),
            'has_secret_config': self.has_secret_config_(),
            'complete_saml_config': self.complete_saml_config_(),
            'complete_oidc_config': self.complete_oidc_config_()
        }